### ISSUE TITLE
Branch 2.0 seems incompatible with version 2.0.1

### ISSUE THREAD

**Issue by dnaber-de** [phpseclib/phpseclib#974]
When comparing the tag `2.0.1` with the branch `2.0`, [a lot of backward incompatible changes seems to be made](https://github.com/phpseclib/phpseclib/compare/2.0.1...2.0#diff-b4b2acaebb5e3d125481ec5da11baff5) to `phpseclib/Crypt/RSA.php`. 

When using composer to define `phpseclib` as dependency (e.g. `phpseclib/phpseclib:~2.0`) with the option `minimum-stability: dev` composer will fetch the branch `2.0` instead of the release `2.0.1` but the branch is no longer compatible with `2.0.1`.

It would be nice if you could make incompatible changes to the API in branches with a new mayor version number, in this case `3.0` and maintain compatibility within the `2.0` branch and releases.

**Comment by terrafrost** [comment]
I accidentally merged the master branch into 2.0 but then I reverted the merge. This was done a few weeks ago. If you have a specific example please provide it but as is I am not aware of anything in the 2.0 branch that shouldn't work identically (baring bug fixes) to how things worked in 2.0.1.

**Comment by dnaber-de** [comment]
Hi terrafrost, I'm using phpseclib via `deployer\deployer`. This is a simplified composer.json that reproduces the problem:

```
{
        "name": "test/test",
        "minimum-stability": "dev",
        "require-dev": {
                "deployer/deployer": "^3.2"
        }
}
```
This resolves to the latest release of deployer [`v3.2.1`](https://github.com/deployphp/deployer/tree/v3.2.1) which depends on phpseclib 2.0: `"phpseclib/phpseclib": "~2.0"`. The minimum-stability option in my composer.json makes composer looking for the `2.0.x-dev` branch and points to b3171cc:

``` 
$ composer info | grep phpsec
phpseclib/phpseclib         2.0.x-dev b3171cc  PHP Secure Communications Library - Pure-PHP implementations of RSA, AES, SSH2, SFTP, X.509 etc.
```
In this commit, there's no method [`phpseclib\Crypt\RSA::loadKey()`](https://github.com/phpseclib/phpseclib/blob/b3171cc63f44c715a2b29200d9549d866e94d123/phpseclib/Crypt/RSA.php) but deployer [depends on it](https://github.com/deployphp/deployer/blob/v3.2.1/src/Server/Remote/PhpSecLib.php#L62). 

To actually trigger an error you'll have to setup a deployment configuration with SSH-Keys like this in the `deploy.php` file:

```
server( 'testing', 'some-host.tld', 22 )
	->user('my_user')
	->identityFile( '~/.ssh/id_rsa.pub', '~/.ssh/id_rsa' )
	->env( 'deploy_path', '/var/www' )
	->env( 'branch', 'master' );
```

**Comment by terrafrost** [comment]
Fuck. I don't know how to fix this. Maybe I can create a branch from the 2.0.1 tag, merge 1.0 into it, copy the files into a temp directory, checkout the 2.0 branch, overwrite the files in the 2.0 branch with the files from the temp directory and then make a commit. #936 would need to be re-applied tho. Others could be as well.

I thought 4b69dcb (a reversion of an earlier merge) fixed the issue since it deleted the RSA subdirectory. After it I merged 2.0 to master and reverted the reversion. But it looks like the reverted reversion is in 2.0 as well FML. And, to top it off, looking more into it, now, I see that 4b69dcb did more than I was wanting anyway as it removed a bunch of entries from CHANGELOG.md. Entries that are present in the 2.0.1 branch.

638fe69 is `Merge branch 'master' of https://github.com/phpseclib/phpseclib` so it seems like that could be the merge that needs to be undone. So I tried `git revert 638fe69 -m 1` and that wasn't it. I did `git reset --hard upstream/2.0` and then did `git revert 638fe69 -m 2` and that's not it either. So I guess it's not that merge.

I don't know what to do :(

**Comment by terrafrost** [comment]
So I went ahead and did what I proposed: #976 is the PR.

The following stuff was manually added (ie. it was in the 2.0 branch but not in the 1.0 branch):

- #936 
- #945 (which was cherry picked from master)
- 9e586e7

This approach doesn't delete files so I deleted those manually:

- phpseclib/Crypt/RSA/*
- phpseclib/Exception/*
- tests/Unit/Crypt/HashTest.php
- tests/Unit/Crypt/RSA/CreateKeyTest.php

**Comment by dnaber-de** [comment]
> I don't know what to do :(

Obviously the problem is not that «popular». This issue will be found by people who run into the same problem, and they'll find how to fix that. `-dev` branches are not suggested to be used in composer anyway. Just keep that issue in mind for the ongoing development. 

Thanks for your effort.

**Comment by terrafrost** [comment]
Looks like I missed #931, which was merged three days after 2.0.1's release. That has now been applied as well.

**Comment by dnaber-de** [comment]
Looks like the problems solved. Thanks a lot :+1:
